python/test/test_binding.py

```diff
@@ -0,0 +1,1152 @@
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+import gc
+import json
+import os
+import queue
+import shutil
+import unittest
+
+import numpy
+from tritonserver import _c as triton_bindings
+
+
+# Callback functions used in inference pipeline
+# 'user_object' is a per-request counter of how many times the
+# callback is invoked
+def g_alloc_fn(allocator, tensor_name, byte_size, memory_type, memory_type_id,
+               user_object):
+    if "alloc" not in user_object:
+        user_object["alloc"] = 0
+    user_object["alloc"] += 1
+    buffer = numpy.empty(byte_size, numpy.byte)
+    return (buffer.ctypes.data, buffer,
+            triton_bindings.TRITONSERVER_MemoryType.CPU, 0)
+
+
+def g_release_fn(allocator, buffer, buffer_user_object, byte_size, memory_type,
+                 memory_type_id):
+    # No-op, buffer ('buffer_user_object') will be garbage collected
+    # only sanity check that the objects are expected
+    if (not isinstance(buffer_user_object, numpy.ndarray)) or (
+            buffer_user_object.ctypes.data != buffer):
+        raise Exception("Misaligned parameters in allocator release callback")
+    pass
+
+
+def g_start_fn(allocator, user_object):
+    if "start" not in user_object:
+        user_object["start"] = 0
+    user_object["start"] += 1
+    pass
+
+
+def g_query_fn(allocator, user_object, tensor_name, byte_size, memory_type,
+               memory_type_id):
+    if "query" not in user_object:
+        user_object["query"] = 0
+    user_object["query"] += 1
+    return (triton_bindings.TRITONSERVER_MemoryType.CPU, 0)
+
+
+def g_buffer_fn(allocator, tensor_name, buffer_attribute, user_object,
+                buffer_user_object):
+    if "buffer" not in user_object:
+        user_object["buffer"] = 0
+    user_object["buffer"] += 1
+    buffer_attribute.memory_type = triton_bindings.TRITONSERVER_MemoryType.CPU
+    buffer_attribute.memory_type_id = 0
+    buffer_attribute.byte_size = buffer_user_object.size
+    return buffer_attribute
+
+
+def g_timestamp_fn(trace, activity, timestamp_ns, user_object):
+    if trace.id not in user_object:
+        user_object[trace.id] = []
+    # not owning trace, so must read property out
+    trace_log = {
+        "id": trace.id,
+        "parent_id": trace.parent_id,
+        "model_name": trace.model_name,
+        "model_version": trace.model_version,
+        "request_id": trace.request_id,
+        "activity": activity,
+        "timestamp": timestamp_ns,
+    }
+    user_object[trace.id].append(trace_log)
+
+
+def g_tensor_fn(
+    trace,
+    activity,
+    tensor_name,
+    data_type,
+    buffer,
+    byte_size,
+    shape,
+    memory_type,
+    memory_type_id,
+    user_object,
+):
+    if trace.id not in user_object:
+        user_object[trace.id] = []
+
+    # not owning trace, so must read property out
```

Comment:
On the current implementation it will be invalid as the py::object is passed as reference to a local variable (wrapper of Triton trace).

To make your example valid, what we can do is to store the trace wrapper in the trace callback resource (`seen_traces`) when seen and remove from it when release.